README.md
  <h3 align=> TRACKER </h3>
   <h3 align=>  BD -TURMA 02-2021 </h3>
   
 <hr>
<p align="center">
  <a href ="#pushpin-objetivos"> Objetivos </a>  • 
  <a href ="#cronograma"> Cronograma </a>  • 
  <a href ="#tecnologias-utilizadas"> Tecnologias </a>  •
  <a href ="#equipe"> Equipe </a>   •
  <a href ="#backlog-total"> Backlog </a>  
 
</p>

<hr>

<br>

# :pushpin: OBJETIVOS

O objetivo é desenvolver um protótipo podendo ser página website ou app, com painel COVID-19 por coleta e Visualização de Dados. Temos as seguintes exigências:

**O protótipo deverá:**

• Utilizar linguagem de programação Python;

• Manipular e tratar dados preferencialmente OFF-LINE (CSV);

• Apresentar o maior número possível de visões contextualizadas (Graficamente ou não);

• Apresentar dados somente do estado de São Paulo;

**O protótipo não poderá:**

• Utilizar 100% de APIs prontas e disponíveis no mercado;

• Utilizar relatórios disponíveis em sites governamentais ou não;

• Não utilizar sistema de Gerenciamento de banco de dado;

<br>

## 📆CRONOGRAMA

   KICK OFF - 16/08 a 22/08; 

   SPRINT 1 - 08/09 a 19/09; 

   SPRINT 2 - 20/09 a 10/10; 

   RECESSO- 11/10 a 17/10;

   SPRINT 3 - 18/10 a  07/11;

   SPRINT 4 - 08/11 a 28/11;

   Apresentação Final - 29/11 a 05/12;

   Feira de Soluções - 16/12;
   
   <br>
   
   **[Sprint 1] - 18/09/2021**
   
   Para o primeiro Sprint, as seguintes adições foram incluídas no programa:
   
   • Cidades registradas no arquivo csv. 
   
   • Filtro de pesquisa por cidade. 
   
   • Número total de mortes. 
   
   • Atualização dos dados de forma automática. 
   
   • reescrita do código para visualização gráfica.

<br>


##  ⚙TECNOLOGIAS UTILIZADAS
<div>
<img src="https://user-images.githubusercontent.com/88864112/133908112-6a9ee071-bc29-4e40-8863-7c5f65caa619.jpg" width="400px" />
<div>

<br>
 
 ## 👥EQUIPE 
 
●   Master

Emanuele Cristina da Silva Diniz Campos

●   Desenvolvedores

Cristiano Alberto Rosa

Flávio Ferreira Antunes

Gabriel Alves Miranda

Gabryel Silva Monteiro

Kelly Carolina Morais Terra

Matheus Aparecido Ramos Lourenço

Renan Moreira Pereira

Wesley Ferreira da Silva
  
<br>
 
 ##  📝BACKLOG TOTAL
  
 <h1 align="left"> <img src = "https://user-images.githubusercontent.com/88864112/133947259-852ee6de-9a47-46c4-9824-351035a4b0e2.jpg" height="650" /></h1>

<br>
 
 <div align="center">
<img src="https://user-images.githubusercontent.com/88864112/133908665-89221a54-c877-430c-a592-74e55414ef0d.jpg"  height="90" /></h1>
<div>
 
 <div align="center"> API 02-2021 - Fatec São José dos Campos


changelog.md

# Change Log
Todas as alterações e melhorias serão encontradas nesse arquivo. 

### Membros do grupo
- Renan Moreira Pereira
-
-
 
## [Sprint 1] - 18/09/2021
 
Para o primeiro Sprint, as seguintes adições foram incluidas no programa:

- Adicionado filtro com todas as cidades registradas no arquivo [csv.](http://brasil.io/datasets)
- Adicionado a possibilidade de digitar o nome da cidade a ser filtrada.
- Incluido o número total de mortes.
- Atualização dos dados de forma automática.
- Reescrito todo o código para futuro suporte de visualização gráfica.


css.py
from PyQt5.uic.properties import QtGui

"""Nesse arquivo se encontram todos os padrões de cores utilizados para botões, backgrounds e etcs."""


class css:
    default = """
                color: #FFFFFF;
                font: 75 10pt "Microsoft YaHei UI";
                """
    blueish = """
                background-color: #336B87;
                """
    blueish_main = """
                InterfacePrincipal {
                background-color: #336B87;
                background-image: url(Xaropinho.png);
                background-position: center;
                background-repeat: no-repeat;
                }
                """
    blueish_text = """
                color: #FFFFFF;
                font-family: Comic Sans MS;
                font-seize: 15px;
                """
    blueish_label = """
                color: #FFFFFF;
                font-family: Comic Sans MS;
                font: bold 15px;
                background-color: #90AFC5;
                border-width: 2px;
                border-radius: 10px;
                font-size: 15px;
                """


update.py
import requests
import gzip
import shutil
import os

# TODO: Alterar o link para ser lido em um arquivo CFG, para evitar futuros problemas.

path = os.path.expanduser('~\Documents\caso_full.csv.gz')
output = os.path.expanduser('~\Documents\caso_full.csv')


def atualizar():
    print("Iniciando o download do arquivo.")
    url = 'https://data.brasil.io/dataset/covid19/caso_full.csv.gz'  # link para o arquivo fornecido pelo site Brasil.IO
    r = requests.get(url, allow_redirects=True)

    with open(path, 'wb') as f:
        f.write(r.content)
        print("Download do arquivo concluido.")
        """Após o download do arquivo, salva na pasta documentos"""

    with gzip.open(path, 'rb') as f_in:
        with open(output, 'wb') as f_out:
            shutil.copyfileobj(f_in, f_out)
        """Extrai o arquivo da fila .GZ na pasta Documentos do usuário"""
        print("Extração concluída.")


interface.py
from PyQt5.QtCore import QRunnable
from matplotlib import pyplot as plt
from covid import tracker
from PyQt5 import QtWidgets
from PyQt5.QtWidgets import *
from datetime import date, timedelta
from css import css
import os
import requests
import shutil
import gzip
import logging
import sys
import threading


# Este arquivo foi criado e modificado por Renan Moreira Pereira (Grupo 4 - FATEC 2021)

# TODO: Adicionar gráficos nas visualizações.
# TODO: Todos os Widgets devem ter a cor alterado ao selecionar o modo escuro.
# TODO: Novas variáveis para melhores visualizações de dados.


class Updating(QRunnable):
    """Função para atualizar os dados do programa."""

    def __init__(self, n):
        super().__init__()

    def atualizar(self):
        logging.info(f"Trabalhando no processo")
        path = os.path.expanduser('~\Documents\caso_full.csv.gz')
        output = os.path.expanduser('~\Documents\caso_full.csv')
        print("Iniciando o download do arquivo.")
        url = 'https://data.brasil.io/dataset/covid19/caso_full.csv.gz'  # link para o arquivo fornecido pelo site Brasil.IO
        r = requests.get(url, allow_redirects=True)

        with open(path, 'wb') as f:
            f.write(r.content)
            logging.info("Main    : Inicializando a thread para atualizar o arquivo.")
            """Após o download do arquivo, salva na pasta documentos"""

        with gzip.open(path, 'rb') as f_in:
            with open(output, 'wb') as f_out:
                shutil.copyfileobj(f_in, f_out)
            """Extrai o arquivo da fila .GZ na pasta Documentos do usuário"""
            logging.info("Main    : Download e extração finalizados.")


class InterfacePrincipal(QMainWindow):
    cidade_selected: str

    def __init__(self):
        super().__init__()

        self.DarkModeStatus = 0

        self.hello = "Seja bem-vindo ao Covid Tracker!"

        self.items = tracker.lista_cidades  # definindo variáveis acima.

        self.select_texto = "Digite o nome de sua cidade:"

        self.data = str(date.today().strftime("%d/%m/%Y"))

        self.data_less_1 = date.today() - timedelta(days=1)
        self.data_string = self.data_less_1.strftime("%y-%m-%d")

        self.data_1 = QtWidgets.QLabel(self.data, self)
        self.data_1.move(950, 520)

        self.atualizarButton = QtWidgets.QPushButton(self)
        self.atualizarButton.clicked.connect(self.atualizar_dados)
        self.atualizarButton.setGeometry(650, 520, 200, 25)
        self.atualizarButton.setText("Atualizar dados.")

        self.grafButton = QtWidgets.QPushButton(self)
        self.grafButton.clicked.connect(self.showGraf)
        self.grafButton.setGeometry(15, 520, 120, 25)
        self.grafButton.setText("Top 10 cidades")

        self.texto = QtWidgets.QLabel(self.hello, self)
        self.texto.setStyleSheet(css.default)
        self.texto.setGeometry(25, 5, 450, 20)

        """Loop criado para inserir as citadel no display inicial"""

        self.lista_header = QtWidgets.QLabel(self.select_texto, self)
        self.lista_header.setGeometry(800, 5, 175, 25)
        self.lista_header.setStyleSheet(css.blueish_text)

        self.lista = QtWidgets.QComboBox(self)
        self.i = 0
        self.j = len(tracker.lista_cidades)

        while self.i < self.j:  # Um loop simples que pega a lista de cidades e coloca num widget de texto.
            self.lista.addItem(tracker.lista_cidades[self.i])
            self.i += 1
        self.lista.setGeometry(800, 35, 165, 30)
        self.lista.setEditable(True)

        self.cidade = self.lista.currentText()

        self.valor_cidade = tracker.df['mortes'].loc[tracker.df['cidade'] == self.cidade].loc[
            tracker.df['tipo'] == 'city'].sum()
        """Filtro se altera com base na cidade que o usuário seleciona"""
        self.valor_dia = tracker.df['mortes'].loc[tracker.df['cidade'] == self.cidade].loc[
            tracker.df['tipo'] == 'city'].loc[tracker.df['data'] == self.data_string].sum()
        """Calcula a data atual - 1 para saber a quantidade de novas mortes """

        self.mortes = QtWidgets.QLabel("   Número total de mortes: " + str(self.valor_cidade), self)
        self.mortes.setStyleSheet(css.blueish_label)
        self.mortes.setGeometry(25, 200, 250, 150)

        self.mortes_dia = QtWidgets.QLabel("   Novas mortes: " + str(self.valor_dia), self)
        self.mortes_dia.setStyleSheet(css.blueish_label)
        self.mortes_dia.setGeometry(25, 300, 250, 150)

        """
        Botões com funções
        """
        self.lista.activated.connect(self.item_usuario)

    def item_usuario(self):
        self.cidade_selected = self.lista.currentText()

        self.valor_cidade_selected = tracker.df['mortes'].loc[tracker.df['cidade'] == self.cidade_selected].loc[
            tracker.df['tipo'] == 'city'].sum()  # Realiza a filtragem de acordo com a cidade selecionada.

        self.mortes.setText("   Número total de Mortes: " + str(int(self.valor_cidade_selected)))
        """Altera o valor dos números de mortes que podemos ver."""

    def atualizar_dados(self):
        formato = "%(asctime)s: %(message)s"
        logging.basicConfig(format=formato, level=logging.INFO, datefmt="%H:%M:%S")
        logging.info("Main    : Inicializando a thread para atualizar o arquivo.")
        x = threading.Thread(target=Updating.atualizar, args=(1,))
        x.start()

    def showGraf(self):
        plt.show()


if __name__ == "__main__":
    app = QApplication(sys.argv)
    app.setStyle('Fusion')
    app.setStyleSheet(css.blueish_main)
    widget = InterfacePrincipal()
    widget.setFixedSize(1024, 550)
    widget.show()
    app.exec()


pandas_exemplo.py
import pandas as pd
import tkinter as tk
import locale

NaN_Value = float("NaN")  # Retira os valores em erro

locale.setlocale(locale.LC_ALL, 'pt_BR.UTF-8')

file = pd.read_csv('caso_full.csv')  # Executa a leitura do arquivo.csv - "file" é apenas o nome que dei a variável.

df = pd.DataFrame(data=file, columns=['city', 'state', 'new_deaths'])  # transforma o arquivo csv em um dataframe
df = df.rename(columns={"city": "Cidade", 'state': 'estado', 'new_deaths': 'mortes'})
df = df.loc[df['estado'] == 'SP']
df = df.append(df.sum(numeric_only=True).rename('Total'))  # Deve ficar próximo do final da manipulação dos dados.
df = df.fillna(value="")

total_mortes = df['mortes'].loc[df['estado'] == 'SP'].sum()  # Calula o total de mortes do estado de SP

print(total_mortes)

cidade_a = str(input("Digite o nome da cidade:"))  # Pega uma cidade, definida pelo usuario, e faz a soma das mortes da mesma.


total_city = df['mortes'].loc[df['Cidade'] == cidade_a].sum()  # Calcula o total de mortes da cidade citada pelo usuário

print(total_city)


def MainWindow():
    front = tk.Tk()  # inicia a interface gráfica.
    front.title = "Covid Tracker"  # Da um nome a janela.
    front.geometry('1060x450')  # Define o tamanho da interface.
    total = tk.Label(front, font=("Courier", 15), text="Total de Mortes: " + str(total_mortes))  # Cria uma caixa de texto
    total.pack()
    cidade = tk.Label(front, font=("Courier", 15), text="Mortes na cidade de " + cidade_a + ":" + str(total_city))
    cidade.pack()
    b = tk.Button(front, text="Ok")
    b.pack()
    front.mainloop()


MainWindow()


covid.py
import locale
import pandas as pd
import os
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import date, timedelta

# 21/09/2021 - Matheus Lourenço - Adicionado gráfico de top 10 cidades mortes por Covid.

locale.setlocale(locale.LC_ALL, 'pt_BR')  # Define a localidade de como o programa deve ser utilizado.


class tracker:
    arquivo = os.path.expanduser('~\Documents\caso_full.csv')  # Local na pasta documentos onde o arquivo estará

    if os.path.isfile(arquivo):
        print("Arquivo encontrado.")
        pass
    else:
        from update import atualizar
        print("Arquivo não encotrado, realizando o download...")
        atualizar()

        """Verifica se o arquivo csv existe, caso a condição seja falsa, irá realizar o download do arquivo."""

    file = pd.read_csv(arquivo)
    """Executa a leitura do arquivo.csv - "file" é apenas o nome que dei a variável."""

    df = pd.DataFrame(data=file,
                      columns=['city', 'date', 'state', 'last_available_deaths', 'place_type',
                               'new_deaths',
                               'new_confirmed'])  # transforma o arquivo csv em um dataframe e seleciona as colunas

    df = df.rename(
        columns={"city": "cidade", 'state': 'estado',
                 'last_available_deaths': 'mortes confirmadas', 'place_type': 'tipo',
                 'new_deaths': 'mortes', 'date': 'data', 'new_confirmed': 'novos casos'})
    """Renomeia as colunas"""

    df = df.loc[df['estado'] == 'SP'].loc[
        df['tipo'] == 'city']
    """Filtra apenas o estado de SP na coluna 'estado', e apenas o cálculo por tipo "state"""

    df = df.fillna(value="")
    """altera os valores de NaN para valor em branco."""

    df = df.drop_duplicates()  # retira valores duplicados.
    df['data'] = pd.to_datetime(df['data'])     # cria uma coluna de data
    df['ano'] = pd.DatetimeIndex(df['data']).year   # cria uma coluna de ano
    df['mes'] = pd.DatetimeIndex(df['data']).month  # cria uma coluna de mes
    df['mes_nome'] = df['data'].dt.strftime('%B')  # transforma o numero da coluna 'mes' para nome do mes
    df['mes/ano'] = df['mes_nome'].astype(str) + "-" + df['ano'].astype(str)    # concatena mes e ano

    '''Variáveis'''

    dia_1 = date.today() - timedelta(days=1)
    dia_1.strftime("%Y-%m-%d")

    total_mortes_sp = df['mortes'].loc[df['estado'] == 'SP'].loc[
        df['data'] == str(dia_1)].sum()  # Soma somente as mortes do estado de SP

    novos_casos = df['novos casos'].loc[df['estado'] == 'SP'].loc[df['data'] == str(dia_1)].sum()  # Soma os casos novos

    total_mortes_dia = df['mortes'].loc[df['data'] == '2020-09-18'].sum()

    lista_cidades = df['cidade'].loc[df['estado'] == 'SP'].loc[df['cidade'] != ''].drop_duplicates().sort_values() \
        .tolist()
    """Transforma os valores em lista, e coloca em ordem alfabética."""

    cidade_sel = df[df.cidade == 'São Paulo']


########################################################################################################################
    sns.set_theme(style="darkgrid")  # faz o gráfico aparecer.

    f, ax = plt.subplots(figsize=(6, 15))

    # gráfico que mostra as 10 maiores cidades por quantidade de mortos
    sns.set_color_codes("pastel")
    sns.barplot(x="mortes", y="cidade", data=df,
                label="Total", color="b", estimator=sum, order=df.value_counts(df['cidade']).iloc[:10].index)
    #plt.show()
